### ISSUE TITLE
integrate(sin(x*cos(x)), x), risch; causes unlimited recursion

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#1183]
Imported from SourceForge on 2024-07-04 01:50:16
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2022-06-01 18:06:33
Original: https://sourceforge.net/p/maxima/bugs/3985

---

The Risch code (src/risch.lisp) has a flag `*IN-RISCH-P*` which is supposed to inhibit reentering RISCHINT when calling MEVAL to evaluate %INTEGRATE nouns in TRIGIN1 and HYPERTRIGINT1. However, it looks that mechanism is not having the expected effect in this case.
```
(%i1) :lisp (trace rischint)
;; Tracing function RISCHINT.
(RISCHINT)
(%i1) integrate(sin(x*cos(x)), x), risch;
1. Trace: (RISCHINT '((%SIN SIMP) ((MTIMES SIMP) $X ((%COS SIMP) $X))) '$X)
2. Trace: 
(RISCHINT
 '((MTIMES SIMP) ((RAT SIMP) -1 2) $%I
   ((MPLUS SIMP)
    ((MTIMES SIMP) -1
     ((MEXPT SIMP) $%E
      ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X
       ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
        ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X))))))
    ((MEXPT SIMP) $%E
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X
      ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
       ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X)))))))
 '$X)
2. Trace: RISCHINT ==> 
((MPLUS SIMP)
 ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I
  ((%INTEGRATE SIMP)
   ((MEXPT SIMP) $%E
    ((MPLUS SIMP)
     ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
     ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
   $X))
 ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I
  ((%INTEGRATE SIMP)
   ((MEXPT SIMP) $%E
    ((MPLUS SIMP)
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
   $X)))
2. Trace: 
(RISCHINT
 '((MEXPT SIMP) $%E
   ((MPLUS SIMP)
    ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
    ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
 '$X)
2. Trace: RISCHINT ==> 
((%INTEGRATE SIMP)
 ((MEXPT SIMP) $%E
  ((MPLUS SIMP)
   ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
   ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
 $X)
2. Trace: 
(RISCHINT
 '((MEXPT SIMP) $%E
   ((MPLUS SIMP)
    ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
    ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
 '$X)
2. Trace: RISCHINT ==> 
((%INTEGRATE SIMP)
 ((MEXPT SIMP) $%E
  ((MPLUS SIMP) ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
   ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
 $X)
2. Trace: (RISCHINT '((%SIN SIMP) ((MTIMES SIMP) $X ((%COS SIMP) $X))) '$X)
3. Trace: 
(RISCHINT
 '((MTIMES SIMP) ((RAT SIMP) -1 2) $%I
   ((MPLUS SIMP)
    ((MTIMES SIMP) -1
     ((MEXPT SIMP) $%E
      ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X
       ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
        ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X))))))
    ((MEXPT SIMP) $%E
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X
      ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
       ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X)))))))
 '$X)
3. Trace: RISCHINT ==> 
[...]
```
I encountered this in a current development version (post-5.46) and it appears to be present also in previous versions -- I tried 5.41 and 5.37.

Somewhat surprising, the nominally-equivalent `risch(sin(x*cos(x)), x)` doesn't run into the looping problem; I see RISCHINT is called twice and then it stops.
```
(%i2) risch(sin(x*cos(x)), x);
1. Trace: (RISCHINT '((%SIN SIMP) ((MTIMES SIMP) $X ((%COS SIMP) $X))) '$X)
2. Trace: 
(RISCHINT
 '((MTIMES SIMP) ((RAT SIMP) -1 2) $%I
   ((MPLUS SIMP)
    ((MTIMES SIMP) -1
     ((MEXPT SIMP) $%E
      ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X
       ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
        ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X))))))
    ((MEXPT SIMP) $%E
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X
      ((MPLUS SIMP) ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X))
       ((MEXPT SIMP) $%E ((MTIMES SIMP) $%I $X)))))))
 '$X)
2. Trace: RISCHINT ==> 
((MPLUS SIMP)
 ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I
  ((%INTEGRATE SIMP)
   ((MEXPT SIMP) $%E
    ((MPLUS SIMP)
     ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
     ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
   $X))
 ((MTIMES SIMP) ((RAT SIMP) -1 2) $%I
  ((%INTEGRATE SIMP)
   ((MEXPT SIMP) $%E
    ((MPLUS SIMP)
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP) -1 $%I $X)))
     ((MTIMES SIMP) ((RAT SIMP) 1 2) $%I $X ((MEXPT SIMP) $%E ((MTIMES SIMP RATSIMP) $%I $X)))))
   $X)))
1. Trace: RISCHINT ==> ((%INTEGRATE SIMP) ((%SIN SIMP) ((MTIMES SIMP) $X ((%COS SIMP) $X))) $X)
                              /
                              [
(%o2)                         I sin(x cos(x)) dx
                              ]
                              /
```
Also, `integrate(sin(x*cos(x)), x)` (without the `risch` flag) calls RISCHINT, but just twice and then it gives up, no looping.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 01:50:18
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2022-06-01 20:27:01
Original: https://sourceforge.net/p/maxima/bugs/3985/#72d3

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 01:50:21
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2022-06-01 20:27:01
Original: https://sourceforge.net/p/maxima/bugs/3985/#b6bc

---

Fixed by commit de20ed5. Closing this report as fixed.
